modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/downsampling/ReplaceBackingWithDownsampleIndexExecutor.java

```diff
@@ -0,0 +1,94 @@
+ * hold up the task listener (nb we notify the listener before we call the delete API so we don't introduce
+ * weird partial failure scenarios - if the delete API fails the
+ * {@link org.elasticsearch.datastreams.lifecycle.DataStreamLifecycleService} will retry on the next run so the source index will get
+ * deleted)
+ */
+public class ReplaceBackingWithDownsampleIndexExecutor extends SimpleBatchedExecutor<ReplaceSourceWithDownsampleIndexTask, Void> {
+    private static final Logger LOGGER = LogManager.getLogger(ReplaceSourceWithDownsampleIndexTask.class);
+    private final Client client;
+
+    public ReplaceBackingWithDownsampleIndexExecutor(Client client) {
+        this.client = client;
+    }
+
+    @Override
+    public Tuple<ClusterState, Void> executeTask(ReplaceSourceWithDownsampleIndexTask task, ClusterState clusterState) throws Exception {
+        return Tuple.tuple(task.execute(clusterState), null);
+    }
+
+    @Override
+    public void taskSucceeded(ReplaceSourceWithDownsampleIndexTask task, Void unused) {
+        LOGGER.trace(
+            "Updated cluster state and replaced index [{}] with index [{}] in data stream [{}]",
+            task.getSourceBackingIndex(),
+            task.getDownsampleIndex(),
+            task.getDataStreamName()
+        );
+        task.getListener().onResponse(null);
+
+        // chain an optimistic delete of the source index call here (if it fails it'll be retried by the data stream lifecycle loop)
+        client.admin().indices().delete(new DeleteIndexRequest(task.getSourceBackingIndex()), new ActionListener<>() {
+            @Override
+            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                LOGGER.info(
+                    "Data stream lifecycle successfully deleted index [{}] due to being replaced by the downsampled index [{}] in"
+                        + " data stream [{}]",
+                    task.getSourceBackingIndex(),
+                    task.getDownsampleIndex(),
+                    task.getDataStreamName()
+                );
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                if (e instanceof IndexNotFoundException) {
+                    // index was already deleted, treat this as a success
+                    return;
+                }
+
+                if (e instanceof SnapshotInProgressException) {
+                    LOGGER.info(
+                        "Data stream lifecycle was unable to delete index [{}] because it's currently being snapshot. Retrying on "
+                            + "the next data stream lifecycle run",
+                        task.getSourceBackingIndex()
+                    );
+                } else {
+                    LOGGER.error(
+                        () -> Strings.format(
+                            "Data stream lifecycle encountered an error trying to delete index [%s]",
```

Comment:
Should this message also indicate that the delete will be retried?

Reply:
++